### PR TITLE
[EPMRPP-38254] Launch log reporting

### PR DESCRIPTION
**Pull request by ihar-kahadouski**

### CODE REVIEW

File: src/main/java/com/epam/reportportal/service/LaunchLoggingContext.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+public class LaunchLoggingContext {
+
+	/* default back-pressure buffer size */
+	public static final int DEFAULT_BUFFER_SIZE = 10;
+	public static final String DEFAULT_LAUNCH_KEY = "default";
+
+	static ConcurrentHashMap<String, LaunchLoggingContext> loggingContextMap = new ConcurrentHashMap<String, LaunchLoggingContext>();
+	/* Log emitter */
+	private final PublishSubject<Maybe<SaveLogRQ>> emitter;
+	/* ID of TestItem in ReportPortal */
```
Comment (pbortnik) [inline src/main/java/com/epam/reportportal/service/LaunchLoggingContext.java]: Is it really test item id?

File: src/main/java/com/epam/reportportal/service/LaunchLoggingContext.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+/**
+ * @author <a href="mailto:ihar_kahadouski@epam.com">Ihar Kahadouski</a>
+ */
+public class LaunchLoggingContext {
+
+	/* default back-pressure buffer size */
+	public static final int DEFAULT_BUFFER_SIZE = 10;
+	public static final String DEFAULT_LAUNCH_KEY = "default";
+
+	static ConcurrentHashMap<String, LaunchLoggingContext> loggingContextMap = new ConcurrentHashMap<String, LaunchLoggingContext>();
```
Comment (pbortnik) [inline src/main/java/com/epam/reportportal/service/LaunchLoggingContext.java]: What is the reason for creating this map?

File: src/main/java/com/epam/reportportal/service/LaunchLoggingContext.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.epam.reportportal.utils.SubscriptionUtils.logFlowableResults;
+import static com.epam.reportportal.utils.files.ImageConverter.convert;
+import static com.epam.reportportal.utils.files.ImageConverter.isImage;
+import static com.google.common.io.ByteSource.wrap;
+
+/**
```
Comment (pbortnik) [inline src/main/java/com/epam/reportportal/service/LaunchLoggingContext.java]: Could you please add documentation as well as in LoggingContext done?

File: src/main/java/com/epam/reportportal/service/ReportPortal.java [hunk 4]
```diff
@@ -191,17 +218,25 @@ public SaveLogRQ apply(String itemId) {
...
+		};
 	}
 
 	public static boolean emitLog(final ReportPortalMessage message, final String level, final Date time) {
-		return emitLog(new com.google.common.base.Function<String, SaveLogRQ>() {
+		return emitLog(getLogSupplier(message, level, time));
+	}
+
+	public static boolean emitLaunchLog(final ReportPortalMessage message, final String level, final Date time) {
+		return emitLaunchLog(getLogSupplier(message, level, time));
```
Comment (pbortnik) [inline src/main/java/com/epam/reportportal/service/ReportPortal.java]: What logSupplier does it use? 
As I see it uses the same as for test item. So you provide item id for it. Is it expected?

File: src/main/java/com/epam/reportportal/service/LaunchLoggingContext.java (hunk 2 above)
Comment (ihar-kahadouski) [reply src/main/java/com/epam/reportportal/service/LaunchLoggingContext.java]: It is possible to start a few launches in parallel. so we need to store context for them

File: src/main/java/com/epam/reportportal/service/LaunchLoggingContext.java (hunk 1 above)
Comment (ihar-kahadouski) [reply src/main/java/com/epam/reportportal/service/LaunchLoggingContext.java]: fixed

File: src/main/java/com/epam/reportportal/service/LaunchLoggingContext.java (hunk 3 above)
Comment (ihar-kahadouski) [reply src/main/java/com/epam/reportportal/service/LaunchLoggingContext.java]: done

**Pull request merged**